sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,752 @@
+
+  return result;
+}
+
+// Send the twin reported property to the service
+static int send_reported_temperature_property(double desired_temp)
+{
+  int rc;
+  printf("Sending reported property\n");
+
+  // Get the topic used to send a reported property update
+  if (az_failed(
+          rc = az_iot_hub_client_twin_patch_get_publish_topic(
+              &client,
+              reported_property_topic_request_id,
+              reported_property_topic,
+              sizeof(reported_property_topic),
+              NULL)))
+  {
+    printf("Unable to get twin document publish topic, return code %d\n", rc);
+    return rc;
+  }
+
+  // Twin reported properties must be in JSON format. The payload is constructed here.
+  az_json_builder json_builder;
+  if (az_failed(rc = build_reported_property(&json_builder, desired_temp)))
+  {
+    return rc;
+  }
+  az_span json_payload = az_json_builder_get_json(&json_builder);
+
+  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+
+  // Publish the reported property payload to IoT Hub
+  if ((rc = MQTTClient_publish(
+           mqtt_client,
+           reported_property_topic,
+           az_span_size(json_payload),
+           az_span_ptr(json_payload),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish reported property, return code %d\n", rc);
+    return rc;
+  }
+  return rc;
+}
+
+// Parse the desired temperature property from the incoming JSON
+static az_result parse_desired_temperature_property(az_span twin_span, double* parsed_value)
+{
+  az_json_parser jp;
+  az_json_token_member tm;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, twin_span));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(desired_temp_property_name, tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm.token, parsed_value));
+    }
+
+    // else ignore token.
+  }
+
+  return AZ_OK;
+}
+
+// Callback for incoming MQTT messages
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  (void)context;
+
+  if (topicLen == 0)
+  {
+    // The length of the topic if there are one or more NULL characters embedded in topicName,
+    // otherwise topicLen is 0.
+    topicLen = (int)strlen(topicName);
+  }
+
+  printf("Topic: %s\n", topicName);
+
+  az_span topic_span = az_span_init((uint8_t*)topicName, topicLen);
+
+  // Parse the incoming message topic and check which feature it is for
+  az_iot_hub_client_twin_response twin_response;
+  az_iot_hub_client_method_request method_request;
+
+  if (az_succeeded(
+          az_iot_hub_client_twin_parse_received_topic(&client, topic_span, &twin_response)))
```

Comment:
This function is +100 lines with a lot going on?  Could we please break it up so that say the block between `az_iot_hub_client_twin_parse_received_topic` is its own function and as would be `az_iot_hub_client_methods_parse_received_topic` block, and this becoems a simple dispatcher whose intent is cleaner?

Also as a nit, the `az_iot_hub_client_methods_parse_received_topic` line way below I think should be an "else if" to indicate that if the `az_iot_hub_client_twin_parse_received_topic` here succeeded we should never be in its path.

Reply:
yes to the `else if` and I will break this up so that in the case that is is either of these, there is some `feature_handler` type thing which does the switch casing or feature specific stuff.